### PR TITLE
AG channel sync; Dutch geodata converter

### PR DESCRIPTION
**Pull request by martijnmartijnmartijnmartijn**
First version of the script that synchronises the AG telegram channels with AN. Requires an .env file with api keys for telegram and AN, links to telegram channels and links to the AN endpoints used. 

Also contains a function to convert Dutch geodata in ag_channels/converter.py - this is probably not the best location, but my Python import skills are lacking, so I couldn't figure out how to place this in its own directory.

### CODE REVIEW

File: ag_channels/sync_ag_channels.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+- links (dict): dictionary mapping the five regions to a telegram invite link
+                for the correponding channels. Example of format:
+        {"Midden":"","Zuid-Oost":"","Zuid-Holland":"","Noord-Oost":"","Noord-West":""}
+- super_admins (list): list of phonenumbers of admins who receive daily logs.
+- an_ag_endpoints (list): list of AN endpoints of the different AG forms.
+
+Requirementss:
+- Telethon           1.16.0
+- phonenumbers       8.12.9
+"""
```
Comment (SamLubbers) [inline ag_channels/sync_ag_channels.py:24]: In Python, the convention is to have the list of requirements in a `requirements.txt` file. This allows any other user to easily installed the required packages with `pip install -r requirements.txt`. [Here](https://www.idkrtm.com/what-is-the-python-requirements-txt/) you can read how to easily create this file.

**Review by SamLubbers — CHANGES_REQUESTED** [review]
@martijnmartijnmartijnmartijn you have the best Github username I've ever seen 🙉 🙉 🙉 🙉 

I have started the review but haven't been able to finish it today. I'll share the existing feedback and requested changes so you can start implementing them in the time it takes me to finish the review.

### Feedback

The code is incredibly well documented and it has a good modular structure, with various concise functions responsible for a single task. This makes the code very easy to understand 💯 

### Requested changes

#### replace global variables with local variables in `converter.py`

Variables used in a single function should always be local variables inside that function, instead of global variables. This makes the code more robust to errors: currently some other code in the application can in import one of these macros and change its value. If that happens the `converter` function would stop working. In contrast, if you move these macros inside the `converter` function, its values cannot be changed by any code outside this function. Here is an example:

```python
def converter(from_data_type, to_data_type, data):
    data_types_index = {
        'postcode': 0,
        'municipality': 1,
        'town': 2,
        'province': 3,
        'region': 4
    }
    try:
        for row in geodata:
            if row[data_types_index[from_data_type]] in data:
                return row[data_types_index[to_data_type]]
    except KeyError:
        raise ValueError(f'from_data_type and to_data_type must have '
                         f'one of the following values: {", ".join(data_types_index.keys())}')
```


Another reason to use local variables is because the code that uses the `converter` function, should only need to import that function to make it work, and not any other variables or functions.

So instead of having this code:

```python
from converter import converter, POSTCODE, MUNICIPALITY, TOWN, PROVINCE, REGION
converter(MUNICIPALITY, REGION, ag["Municipality"])
```
You would have the following code

```python
from converter import converter
converter("municipality", "region", ag["Municipality"])
```

Of course, you would need to update the documentation to explain what are the possible values for `from_data_type` and `to_data_type`:

```python
def converter(from_data_type, to_data_type, data):
    """Here you write what the function does

    from_data_type: string indicating the type of the data variable. Allowed values are: postcode, municipality, town, province, region
    to_data_type: ...
    data: ...
    """
```

File: ag_channels/converter.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        would call:
+            "convert(POSTCODE, MUNICIPALITY, '1098AB')"
+
+        TODO: filter nonsensical combination of macros, e.g.:
+            convert(TOWN, POSTCODE, 'xxx')
+        TODO: filter invalid input data for each input type.
+    """
+    for row in geodata:
+        if row[from_data_type] in data:
+            return row[to_data_type]
```
[the remark this answers is not in the document]
Comment (martijnmartijnmartijnmartijn) [reply ag_channels/converter.py]: Yeah I agree that those are better solutions, but for now the conversion is nearly instant. I've already run the long script to convert all our members' egodata, so for now we will only use this function for individual cases, for which, as you already said, it should be good enough.

**Review by SamLubbers — APPROVED** [review]
It's a beautiful piece of art

**Pull request merged**